pkg/validation-webhook/snapshot.go

```diff
@@ -0,0 +1,245 @@
+		}
+		return decideSnapshotContent(snapcontent, oldSnapcontent, isUpdate)
+	default:
+		err := fmt.Errorf("expect resource to be %s or %s", SnapshotGVR, SnapshotContentGVR)
+		klog.Error(err)
+		return toV1AdmissionResponse(err)
+	}
+}
+
+func decideSnapshot(snapshot, oldSnapshot *volumesnapshotv1beta1.VolumeSnapshot, isUpdate bool) *v1.AdmissionResponse {
+	reviewResponse := v1.AdmissionResponse{}
+	reviewResponse.Allowed = true
+	reviewResponse.Result = &metav1.Status{}
+
+	// if it is an UPDATE and oldSnapshot is not valid, then don't enforce strict validation
+	// This allows no-op updates to occur on snapshot resources which fail strict validation
+	// Which allows the remover of finalizers and therefore deletion of this object
+	err := IsValidSnapshot(oldSnapshot)
+	if isUpdate && err != nil {
+		return &reviewResponse
+	}
+
+	err = IsValidSnapshot(snapshot)
+	if err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return &reviewResponse
+	}
+
+	err = checkSnapshotImmutableFields(snapshot, oldSnapshot)
+	if isUpdate && err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return &reviewResponse
+	}
+
+	return &reviewResponse
+}
+
+func decideSnapshotContent(snapcontent, oldSnapcontent *volumesnapshotv1beta1.VolumeSnapshotContent, isUpdate bool) *v1.AdmissionResponse {
+	reviewResponse := v1.AdmissionResponse{}
+	reviewResponse.Allowed = true
+	reviewResponse.Result = &metav1.Status{}
+
+	// if it is an UPDATE and oldSnapshot is not valid, then don't enforce strict validation
+	// This allows no-op updates to occur on snapshot resources which fail strict validation
+	// Which allows the remover of finalizers and therefore deletion of this object
+	err := IsValidSnapshotContent(oldSnapcontent)
+	if isUpdate && err != nil {
+		return &reviewResponse
+	}
+
+	err = IsValidSnapshotContent(snapcontent)
+	if err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return &reviewResponse
+	}
+
+	err = checkSnapshotContentImmutableFields(snapcontent, oldSnapcontent)
+	if isUpdate && err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+		return &reviewResponse
+	}
+
+	return &reviewResponse
+}
+
+// IsValidSnapshot performs additional strict validation.
+// Do NOT rely on this function to fully validate snapshot objects.
+// This function will only check the additional rules provided by the webhook.
+func IsValidSnapshot(snapshot *volumesnapshotv1beta1.VolumeSnapshot) error {
+	if snapshot == nil {
+		return fmt.Errorf("Snapshot is nil")
+	}
+
+	source := snapshot.Spec.Source
+
+	if source.PersistentVolumeClaimName != nil && source.VolumeSnapshotContentName != nil {
+		return fmt.Errorf("Only one of Spec.Source.PersistentVolumeClaimName = %s and Spec.Source.VolumeSnapshotContentName = %s should be set.", *source.PersistentVolumeClaimName, *source.VolumeSnapshotContentName)
```

Comment:
Usually errors should start with lower-case letter and shouldn't end with a dot, same below.
https://github.com/golang/go/wiki/CodeReviewComments#error-strings